Prevent the git_trace callback from being garbage collected

jeffhostetler opened the pull request:
Here is a better fix for the problem with the git_trace_cb being GC'd.

Thanks to @jamill for the suggestion that the call in GlobalSettings to Proxy.git_trace_set() might be secretly creating a temporary delegate for the GitTraceHandler method.

Therzok commented:
This is way better, thanks!

nulltoken commented:
@jeffhostetler Neat!

Few things though:

 - It looks like the identity you used to craft your commit isn't known from GitHub. As such, there's no link between the commit and your profile. You might be willing to fix that.
 - The commit message actually relates to a discussion that happened in #915. Could you please  reword it so that it makes sense by itself? Something as simple as "Prevent the git_trace callback from being garbage collected" would do the trick.

jeffhostetler commented:
Sorry, I pushed that from my office machine.  I've added that email address to my account.

I've rebased and reworded the commit as suggested.  Thanks for the feedback.

jamill commented on LibGit2Sharp/Core/Proxy.cs:
```diff
@@ -2959,6 +2959,18 @@ public static GitObjectType git_tag_target_type(GitObjectSafeHandle tag)
+        ///
+        /// Since the given callback will be passed into and retained by C code,
+        /// it is very important that you pass an actual delegate here (and don't
+        /// the compiler create/cast a temporary one for you).  Furthermore, you
```
and don't [let] the compiler...?

nulltoken commented:
@jeffhostetler Thanks a lot :sparkles: 

As you're at ease with rebasing, could you please rebase onto the latest vNext once @jamill's comment is fixed?

jeffhostetler commented:
Fixed. Thanks.

nulltoken commented:
Awesome. Just a minor warning from the build log:

```
"C:\projects\libgit2sharp\LibGit2Sharp.sln" (default target) (1) ->
"C:\projects\libgit2sharp\LibGit2Sharp\LibGit2Sharp.csproj" (default target) (2) ->
(CoreCompile target) -> 
  Core\Proxy.cs(2973,9): warning CS1570: XML comment on 'LibGit2Sharp.Core.Proxy.git_trace_set(LibGit2Sharp.LogLevel, LibGit2Sharp.Core.NativeMethods.git_trace_cb)' has badly formed XML -- 'End tag 'member' does not match the start tag 'summary'.' [C:\projects\libgit2sharp\LibGit2Sharp\LibGit2Sharp.csproj]
```

jeffhostetler commented:
Sorry about that.  Got it.

nulltoken commented:
:sparkles: :sparkles: :sparkles: :sparkles: :sparkles: :sparkles: :bangbang:
